File: content/tracks/c/concepts/02-functions/functions_test.c
/*
 * Build: gcc -o test stub.c functions_test.c && ./test
 * (Remove or guard the main() in stub.c with #ifndef TESTING before linking.)
 */
#include <stdio.h>

int max(int a, int b);
int factorial(int n);
int is_palindrome(int n);

static int passed = 0, failed = 0;

#define CHECK_INT(name, expected, actual) \
    do { \
        int _e = (expected), _a = (actual); \
        if (_e == _a) { printf("  PASS: %s\n", name); passed++; } \
        else { printf("  FAIL: %s — expected %d, got %d\n", name, _e, _a); failed++; } \
    } while (0)

int main(void) {
    CHECK_INT("max(3,7)",   7, max(3, 7));
    CHECK_INT("max(7,3)",   7, max(7, 3));
    CHECK_INT("max(5,5)",   5, max(5, 5));
    CHECK_INT("max(-1,-2)", -1, max(-1, -2));

    CHECK_INT("factorial(0)",  1,   factorial(0));
    CHECK_INT("factorial(1)",  1,   factorial(1));
    CHECK_INT("factorial(5)",  120, factorial(5));
    CHECK_INT("factorial(7)",  5040, factorial(7));

    CHECK_INT("is_palindrome(0)",    1, is_palindrome(0));
    CHECK_INT("is_palindrome(1)",    1, is_palindrome(1));
    CHECK_INT("is_palindrome(121)",  1, is_palindrome(121));
    CHECK_INT("is_palindrome(1221)", 1, is_palindrome(1221));
    CHECK_INT("is_palindrome(123)",  0, is_palindrome(123));
    CHECK_INT("is_palindrome(120)",  0, is_palindrome(120));

    printf("\n%d passed, %d failed\n", passed, failed);
    return failed > 0 ? 1 : 0;
}

File: content/tracks/kotlin/concepts/02-functions/functions_test.kt
var passed = 0
var failed = 0

fun test(name: String, block: () -> Unit) {
    try {
        block()
        println("  PASS: $name")
        passed++
    } catch (e: Exception) {
        println("  FAIL: $name — ${e.message}")
        failed++
    }
}

fun assertEquals(expected: Any?, actual: Any?, msg: String = "") {
    if (expected != actual) throw Exception("expected $expected but got $actual${if (msg.isNotEmpty()) " ($msg)" else ""}")
}

fun main() {
    println("repeat")
    test("default times=2") { assertEquals("haha", repeat("ha")) }
    test("explicit times") { assertEquals("ababab", repeat("ab", 3)) }
    test("times=1") { assertEquals("x", repeat("x", 1)) }
    test("times=0") { assertEquals("", repeat("x", 0)) }

    println("wordCount")
    test("two words") { assertEquals(2, "hello world".wordCount()) }
    test("leading/trailing spaces") { assertEquals(2, "  foo bar  ".wordCount()) }
    test("multiple spaces between") { assertEquals(3, "a  b   c".wordCount()) }
    test("empty string") { assertEquals(0, "".wordCount()) }
    test("single word") { assertEquals(1, "kotlin".wordCount()) }

    println("joinWith")
    test("default separator") { assertEquals("a, b, c", joinWith("a", "b", "c")) }
    test("custom separator") { assertEquals("a|b|c", joinWith("a", "b", "c", separator = "|")) }
    test("single part") { assertEquals("only", joinWith("only")) }
    test("empty string parts") { assertEquals(", ", joinWith("", "")) }

    println("pow")
    test("2 pow 10") { assertEquals(1024, 2 pow 10) }
    test("5 pow 0") { assertEquals(1, 5 pow 0) }
    test("3 pow 3") { assertEquals(27, 3 pow 3) }
    test("1 pow 100") { assertEquals(1, 1 pow 100) }
    test("10 pow 3") { assertEquals(1000, 10 pow 3) }

    println("\n$passed passed, $failed failed")
    if (failed > 0) kotlin.system.exitProcess(1)
}
